google/cloud/storage/transfer_manager.py

```diff
@@ -495,7 +702,226 @@ def download_many_to_path(
+
+    :type worker_type: str
+    :param worker_type:
+        The worker type to use; one of google.cloud.storage.transfer_manager.PROCESS
+        or google.cloud.storage.transfer_manager.THREAD.
+
+        Although the exact performance impact depends on the use case, in most
+        situations the PROCESS worker type will use more system resources (both
+        memory and CPU) and result in faster operations than THREAD workers.
+
+        Because the subprocesses of the PROCESS worker type can't access memory
+        from the main process, Client objects have to be serialized and then
+        recreated in each subprocess. The serialization of the Client object
+        for use in subprocesses is an approximation and may not capture every
+        detail of the Client object, especially if the Client was modified after
+        its initial creation or if `Client._http` was modified in any way.
+
+        THREAD worker types are observed to be relatively efficient for
+        operations with many small files, but not for operations with large
+        files. PROCESS workers are recommended for large file operations.
+
+    :type max_workers: int
+    :param max_workers:
+        The maximum number of workers to create to handle the workload.
+
+        With PROCESS workers, a larger number of workers will consume more
+        system resources (memory and CPU) at once.
+
+        How many workers is optimal depends heavily on the specific use case,
+        and the default is a conservative number that should work okay in most
+        cases without consuming excessive resources.
+
+    :raises: :exc:`concurrent.futures.TimeoutError` if deadline is exceeded.
+    """
+
+    if download_kwargs is None:
+        download_kwargs = {}
+    forced_start = download_kwargs.pop("start", 0)
+    forced_end = download_kwargs.pop("end", None)
+    # We must know the size of the object, and the generation.
+    if not blob.size or not blob.generation:
+        blob.reload()
+
+    pool_class, needs_pickling = _get_pool_class_and_requirements(worker_type)
+    # Pickle the blob ahead of time (just once, not once per chunk) if needed.
+    maybe_pickled_blob = _pickle_blob(blob) if needs_pickling else blob
+
+    futures = []
+
+    # Create and/or truncate the destination file to prepare for sparse writing.
+    with open(filename, "wb") as _:
+        pass
+
+    with pool_class(max_workers=max_workers) as executor:
+        cursor = forced_start
+        # forced_end is zero-indexed here, so add 1
+        end = min(forced_end + 1, blob.size) if forced_end else blob.size
+        while cursor < end:
+            start = cursor
+            cursor = min(cursor + chunk_size, end)
+            futures.append(
+                executor.submit(
+                    _download_and_write_chunk_in_place,
+                    maybe_pickled_blob,
+                    filename,
+                    offset=(forced_start * -1),
```

Comment:
Could you help me understand how the offset is computed here?

Reply:
Might be deleting this anyways if we decide ranged reads are out of scope. If not, I should document this entire section better as there are a couple of tricky spots.